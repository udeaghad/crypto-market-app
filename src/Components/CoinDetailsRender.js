import { useSelector } from 'react-redux';
import React from 'react';
import './coinPage.css';

const DetailsRender = () => {
  const coinExchange = useSelector((state) => state.cryptoMarkets);
  const exchangeMarket = coinExchange.market;

  const coinExchangeId = coinExchange.coin_id;

  const getCoinDetails = useSelector((state) => state.cryptoCoins);

  const coinDetails = getCoinDetails.coin;
  let currency = 'USD';
  if (getCoinDetails.currency) {
    currency = getCoinDetails.currency;
  }

  let newDetail = [{ name: 'bitcoin', price: 0 }];
  if (coinDetails) {
    newDetail = coinDetails.filter((item) => item.id === coinExchangeId);
  }

  return (
    <div className="pg-container">
      {newDetail.length
        ? (
          <div className="coin-head-pg">
            <div className="coin-photo">
              <img className="icon-pix" src={newDetail[0].icon} alt="" />
            </div>
            <div className="coin-name-price">
              <h2 className="coin-pg-name">{newDetail[0].name}</h2>
              <div className="coin-pg-price">{newDetail[0].price.toLocaleString('en-US', { style: 'currency', currency })}</div>
            </div>
          </div>
        )
        : <div>Loading...</div>}

      <div className="breakdown-label">
        <h4>Exchange Breakdown</h4>
      </div>

      <div className="coin-pg-exchange">
        {exchangeMarket
          ? exchangeMarket.map((coin, index) => (
            <div className="coin-exchange" key={coin.id} style={{ backgroundColor: index % 2 !== 0 ? 'rgb(76 102 146 / 1)' : null }}>
              <div className="exchange-name">
                <p>{coin.exchange}</p>
              </div>
              <div className="exchange-details">
                <p>
                  Pair:
                  {' '}
                  {coin.pair}
                </p>
                <p>
                  Pair Price:
                  {' '}
                  {coin.pairPrice}
                </p>
                <p>
                  Price:
                  {' '}
                  {coin.price}
                </p>
                <p>
                  Volume:
                  {' '}
                  {coin.volume}
                </p>
              </div>

            </div>
          ))
          : <p>Loading...</p>}
      </div>

    </div>
  );
};

export default DetailsRender;
