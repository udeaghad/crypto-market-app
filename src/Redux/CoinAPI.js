import { createAsyncThunk } from '@reduxjs/toolkit';


const GET_ALL_COINS = 'GET_ALL_COINS'
const GET_COIN_MARKETS = 'GET_COIN_MARKETS'

const baseUrl = 'https://api.coinstats.app/public/v1';

const getAllCoins = createAsyncThunk(
  GET_ALL_COINS,
  async () => {
    const request = new Request(`${baseUrl}/coins?skip=0&limit=10&currency=USD`);
    const response = await fetch(request);
    const result = await response.json();
   
    return result.coins
  }
)

const getCoinMarkets = createAsyncThunk(
  GET_COIN_MARKETS,
  async (coinId) => { 
    console.log(coinId) 
    const request = new Request(`${baseUrl}/markets?coinId=${coinId}`);
    const response = await fetch(request);
    const result = await response.json();
    

    let markets =[];
    for (let i =0; i < 10; i+=1){
      markets.push({...result[i], id:i+1})
    }
   
    return ({market: markets, coin_id: coinId})
    
  }
)

export {getAllCoins, getCoinMarkets};