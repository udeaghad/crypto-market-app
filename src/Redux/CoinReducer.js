const cryptoMarket = [];

let newState

const getMarketReducer = (state = cryptoMarket, action) => {
  switch (action.type) {
    case 'GET_COIN_MARKETS/fulfilled':
      return action.payload;
    default:
      return state;
  }
};
const cryptoCoins = [];

const getCoinsReducer = (state = cryptoCoins, action) => {
  switch (action.type) {
    case 'GET_ALL_COINS/fulfilled':
      return action.payload;
    
    case 'GET_BY_RANK':
      console.log(action.payload)
      if(action){
        newState = state.coin.filter(item => item.rank <= action.payload)
        return {...state, coin:newState}
      }
      
    default:
      return state;
  }
};

export { getMarketReducer, getCoinsReducer };
