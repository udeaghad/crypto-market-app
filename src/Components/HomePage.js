import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { NavLink } from 'react-router-dom';
import { MdSettings } from 'react-icons/md';
import { FaMicrophone } from 'react-icons/fa';
import { BsArrowRightCircle } from 'react-icons/bs';
import { getAllCoins, getCoinMarkets, searchByRank } from '../Redux/CoinAPI';
import cryptoPix from './images/crypto-pix.jpg';

const HomePage = () => {

  const getList = useSelector((state) => state.cryptoCoins);

  const coins = getList.coin;
  const { currency } = getList;

  const dispatch = useDispatch();

  let cur = 'USD';

  const onSelect = (e) => {    
     cur = e.target.value     
    dispatch(getAllCoins(cur));
    let parent = e.target.parentElement.children[1]
    parent.children[0].value = '';
  };

  useEffect(() => {      
      dispatch(getAllCoins(cur));     
         
  }, []);

  let rank = '';

  const onFillValue =(e)=> {
    rank = e.target.value
    
   }

  const onSearch = (e) => {
   e.preventDefault();   
   
    if(rank){
      dispatch(searchByRank(rank))
      
    }else if(!rank){
      dispatch(getAllCoins(cur))
   
        }     
    
  }

  const onClear = (e) => {   
    dispatch(getAllCoins(cur))
    e.target.parentElement.firstElementChild.value = ""
  }

  const onBackspace = e => {
    if(e.key === 'Backspace'){
      dispatch(getAllCoins(cur))
    }
  }  

  const handleClick = (e) => {
    dispatch(getCoinMarkets(e.target.id));
  };

  return (
    <div className="pg-container">
      <nav>
        <p className="heading">Top Cryptos</p>
        <div className="nav-icon">
          <FaMicrophone style={{ color: 'white', marginRight: '2rem' }} />
          <MdSettings style={{ color: 'white' }} />
        </div>
      </nav>
      <div className="pg-head">
        <div className="crypto-pix">
          <img className="image" src={cryptoPix} alt="currency" />
        </div>
        <div className="head-desr">
          <h1 className="crypto-x">CRYPTO EXCHANGE</h1>
        </div>

      </div>

      <div className="coin-heading">
        <div>
        <h4>Price Per Coin</h4> 
        </div>
          

        <div className='search-row'>

       
        <select
          className="currency-name"
          name="currency"
          type="text"
          onChange={(e) => onSelect(e)}          
        >
          <option value="USD">Choose Currency</option>
          <option value="USD">USD - US Dollars</option>
          <option value="EUR">EUR - EURO</option>
          <option value="GBP">GBP - British Pounds</option>
          <option value="CAD">CAD - Canadian Dollars</option>
          <option value="AUD">AUD - Australian Dollars</option>
        </select>

        <form onSubmit={(e) => onSearch(e)}>
        <input
        name='search'
        type='number'
        className='search-field'
        placeholder='Search Top Rank'
        onChange={(e) => onFillValue(e)}
        onKeyDown={e => onBackspace(e)} 
        max='50'       
        />
        <button className='filter-btn' type='submit'>Filter</button>
        <button className='clear-btn' type='button' onClick={(e) => onClear(e)}>Clear</button>
        </form>
        </div>
      </div>
      <div className="list-coins">
     
        {coins ? coins.map((unitCoin) => (
         
          <div
            key={unitCoin.id}
            className="unit-coin"
          >
            <NavLink to="details" onClick={(e) => handleClick(e)} className="arrow-icon">
              <div>
                <BsArrowRightCircle id={unitCoin.id} style={{ width: '20px', height: '20px', color: 'white' }} />
              </div>
            </NavLink>

            <img src={unitCoin.icon} alt="crypto-coin" className="crypto-coin" />
            <div className="coin-name">{unitCoin.name}</div>
            <div className='price-rank'>
            <div className="coin-price">
              {unitCoin.price.toLocaleString('en-US', { style: 'currency', currency })}
            </div>
            <div className='coin-rank'>Rank: {unitCoin.rank}</div>
            </div>
          </div>

        ))
          : (
            <div className="loading-page">
              <p>Loading...wait for some seconds</p>
            </div>
          )}

      </div>
    </div>

  );
};

export default HomePage;
