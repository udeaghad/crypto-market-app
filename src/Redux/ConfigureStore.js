import { combineReducers, configureStore } from '@reduxjs/toolkit';
import { getCoinsReducer, getMarketReducer } from './CoinReducer';

const rootReducer = combineReducers({
  cryptoCoins: getCoinsReducer,
  cryptoMarkets: getMarketReducer,
});

const store = configureStore({
  reducer: rootReducer,
});

export default store;
