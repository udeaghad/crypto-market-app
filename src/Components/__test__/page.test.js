import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import store from '../../Redux/ConfigureStore';
import HomePage from '../HomePage';
import CoinDetailsRender from '../CoinDetailsRender';
import { getCoinsReducer } from '../../Redux/CoinReducer';
import { searchByRank } from '../../Redux/CoinAPI';

describe('To take snapshot of all components', () => {
  test('To Test HomePage component if renders successfully!', () => {
    const tree = renderer
      .create(<Provider store={store}><HomePage /></Provider>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('To Test CoinDetailRender component if renders successfully!', () => {
    const tree = renderer
      .create(<Provider store={store}><CoinDetailsRender /></Provider>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('To test the filter function of the reducer', () => {
    const state = {
      coin: [
        {
          id: 'binance', name: 'Bitcoin', rank: 1, price: 23000,
        },
        {
          id: 'ethereum', name: 'Ethereum', rank: 2, price: 1800,
        },
        {
          id: 'tether', name: 'Tether', rank: 3, price: 1,
        },
        {
          id: 'bnb', name: 'BNB', rank: 4, price: 307,
        },
        {
          id: 'usd_coin', name: 'USD Coin', rank: 5, price: 0.8,
        },
      ],
      currency: 'USD',

    };

    expect(getCoinsReducer(state, searchByRank(3))).toEqual(
      {
        coin: [
          {
            id: 'binance', name: 'Bitcoin', rank: 1, price: 23000,
          },
          {
            id: 'ethereum', name: 'Ethereum', rank: 2, price: 1800,
          },
          {
            id: 'tether', name: 'Tether', rank: 3, price: 1,
          },
        ],
        currency: 'USD',
      },

    );
  });
});
