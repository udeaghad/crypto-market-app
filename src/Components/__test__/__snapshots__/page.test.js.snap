// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`To take snapshot of all components To Test CoinDetailRender component if renders successfully! 1`] = `
<div
  className="pg-container"
>
  <div
    className="coin-head-pg"
  >
    <div
      className="coin-photo"
    >
      <img
        alt=""
        className="icon-pix"
      />
    </div>
    <div
      className="coin-name-price"
    >
      <h2
        className="coin-pg-name"
      >
        bitcoin
      </h2>
      <div
        className="coin-pg-price"
      >
        $0.00
      </div>
    </div>
  </div>
  <div
    className="breakdown-label"
  >
    <h4>
      Exchange Breakdown
    </h4>
  </div>
  <div
    className="coin-pg-exchange"
  >
    <p>
      Loading...
    </p>
  </div>
</div>
`;

exports[`To take snapshot of all components To Test HomePage component if renders successfully! 1`] = `
<div
  className="pg-container"
>
  <nav>
    <p
      className="heading"
    >
      Top Cryptos
    </p>
    <div
      className="nav-icon"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": "white",
            "marginRight": "2rem",
          }
        }
        viewBox="0 0 352 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M176 352c53.02 0 96-42.98 96-96V96c0-53.02-42.98-96-96-96S80 42.98 80 96v160c0 53.02 42.98 96 96 96zm160-160h-16c-8.84 0-16 7.16-16 16v48c0 74.8-64.49 134.82-140.79 127.38C96.71 376.89 48 317.11 48 250.3V208c0-8.84-7.16-16-16-16H16c-8.84 0-16 7.16-16 16v40.16c0 89.64 63.97 169.55 152 181.69V464H96c-8.84 0-16 7.16-16 16v16c0 8.84 7.16 16 16 16h160c8.84 0 16-7.16 16-16v-16c0-8.84-7.16-16-16-16h-56v-33.77C285.71 418.47 352 344.9 352 256v-48c0-8.84-7.16-16-16-16z"
        />
      </svg>
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": "white",
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M19.14 12.94c.04-.3.06-.61.06-.94 0-.32-.02-.64-.07-.94l2.03-1.58a.49.49 0 00.12-.61l-1.92-3.32a.488.488 0 00-.59-.22l-2.39.96c-.5-.38-1.03-.7-1.62-.94l-.36-2.54a.484.484 0 00-.48-.41h-3.84c-.24 0-.43.17-.47.41l-.36 2.54c-.59.24-1.13.57-1.62.94l-2.39-.96c-.22-.08-.47 0-.59.22L2.74 8.87c-.12.21-.08.47.12.61l2.03 1.58c-.05.3-.09.63-.09.94s.02.64.07.94l-2.03 1.58a.49.49 0 00-.12.61l1.92 3.32c.12.22.37.29.59.22l2.39-.96c.5.38 1.03.7 1.62.94l.36 2.54c.05.24.24.41.48.41h3.84c.24 0 .44-.17.47-.41l.36-2.54c.59-.24 1.13-.56 1.62-.94l2.39.96c.22.08.47 0 .59-.22l1.92-3.32c.12-.22.07-.47-.12-.61l-2.01-1.58zM12 15.6c-1.98 0-3.6-1.62-3.6-3.6s1.62-3.6 3.6-3.6 3.6 1.62 3.6 3.6-1.62 3.6-3.6 3.6z"
        />
      </svg>
    </div>
  </nav>
  <div
    className="pg-head"
  >
    <div
      className="crypto-pix"
    >
      <img
        alt="currency"
        className="image"
        src="crypto-pix.jpg"
      />
    </div>
    <div
      className="head-desr"
    >
      <h1
        className="crypto-x"
      >
        CRYPTO EXCHANGE
      </h1>
    </div>
  </div>
  <div
    className="coin-heading"
  >
    <div>
      <h4>
        Price Per Coin
      </h4>
    </div>
    <div
      className="search-row"
    >
      <select
        className="currency-name"
        name="currency"
        onChange={[Function]}
        type="text"
      >
        <option
          value="USD"
        >
          Choose Currency
        </option>
        <option
          value="USD"
        >
          USD - US Dollars
        </option>
        <option
          value="EUR"
        >
          EUR - EURO
        </option>
        <option
          value="GBP"
        >
          GBP - British Pounds
        </option>
        <option
          value="CAD"
        >
          CAD - Canadian Dollars
        </option>
        <option
          value="AUD"
        >
          AUD - Australian Dollars
        </option>
      </select>
      <form
        onSubmit={[Function]}
      >
        <input
          className="search-field"
          max="50"
          name="search"
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="Search Top Rank"
          type="number"
        />
        <button
          className="filter-btn"
          type="submit"
        >
          Filter
        </button>
        <button
          className="clear-btn"
          onClick={[Function]}
          type="button"
        >
          Clear
        </button>
      </form>
    </div>
  </div>
  <div
    className="list-coins"
  >
    <div
      className="loading-page"
    >
      <p>
        Loading...wait for some seconds
      </p>
    </div>
  </div>
</div>
`;
